### PR TITLE
Customize your store header

### LINKED ISSUE #39718: [Customize Your Store] Header
## User Story

> As a new merchant, I would like a to fine tune the AI design, so I can add my personal touch.

## Design

* [Figma](https://www.figma.com/file/Mrk6SERPZ4KrFHSjM0a8TK/Customize-your-store-with-pattern-assembler?node-id=2631%3A255284&mode=dev)

Image

## Dev

**Environment setup**

- Use WP 6.3
- Install the WooCommerce Beta Tester plugin
- Go to /wp-admin/tools.php?page=woocommerce-admin-test-helper and enable customize-store feature flag
- Go to WooCommerce > Home
- Click on `Customize your store` task
- Click on `Assembler Hub` button
- YousShould see the Assembler Hub screen

**Implement the Header sidebar screen**

Will need to update [this header screen component](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce-admin/client/customize-store/assembler-hub/sidebar/sidebar-navigation-screen-header.tsx#L31-L35
): 

1. Regarding how to pull pattern data, please refer to [these lines](https://github.com/woocommerce/woocommerce/pull/38765/files#diff-f4e9a0bbe4f91e230f73dc942471f76d9863327fbfd43c567831f679d84af9c6R35-R61) and we'll need to filter patterns by:
  - categories: "woo-commerce"
  - blockTypes : `core/template-part/header`

2. Then, we can use the [BlockPatternList](https://github.com/WordPress/gutenberg/blob/ecbde0e846ce3958f125302cfeda1cd34b5b89e3/packages/block-editor/src/components/block-patterns-list) component to show the patterns on the sidebar. [example](https://github.com/woocommerce/woocommerce/pull/38765/files#diff-f4e9a0bbe4f91e230f73dc942471f76d9863327fbfd43c567831f679d84af9c6R138-R146). Every WC header pattern have the `wc-blocks-header-pattern` class name, we could use it to add opacity styles.

3. To replace the header template part, first we'll need to find the current header block's client ID. 

https://github.com/woocommerce/woocommerce/blob/914df4b921acc91c6ad8612449216e18d49ac0a5/plugins/woocommerce-admin/client/customize-store/assembler-hub/block-editor.tsx#L50C1-L53C4

A block object would be like

```json
{
    "clientId": "79d4bd3d-4a9d-470d-aca5-0510d9a11b85",
    "name": "core/group",
    "isValid": true,
    "attributes": {
        "tagName": "div",
        "align": "full",
        "style": {
            "spacing": {
                "blockGap": "0px",
                "padding": {
                    "top": "1rem",
                    "right": "1rem",
                    "bottom": "1rem",
                    "left": "1rem"
                }
            },
            "elements": {
                "link": {
                    "color": {
                        "text": "var:preset|color|background"
                    }
                }
            }
        },
        "backgroundColor": "black",
        "textColor": "white",
        "className": "wc-blocks-header-pattern has-background-color",
        "layout": {
            "type": "default"
        }
    },
    "innerBlocks": [
        {
            "clientId": "2aef703e-6201-45d1-9517-d6e9039887ab",
            "name": "core/group",
            "isValid": true,
            "originalContent": "<div class=\"wp-block-group has-small-font-size\">\n\t\t\n\n\t\t\n</div>",
            "validationIssues": [],
            "attributes": {
                "tagName": "div",
                "style": {
                    "spacing": {
                        "blockGap": "1rem"
                    }
                },
                "className": "has-small-font-size",
                "layout": {
                    "type": "flex",
                    "flexWrap": "nowrap",
                    "justifyContent": "space-between"
                }
            },
        }
     ]
}
```

Then, we can use [`updateBlockAttributes`](https://github.com/WordPress/gutenberg/blob/6b1e9377937b7c3ad1704e4979179700d5ba456a/packages/block-editor/src/store/actions.js#L159) action to update the block attributes. Please refer [here](https://github.com/WordPress/gutenberg/blob/831a536c8ca60d8d8d84380052761a4609c4c1d4/packages/block-library/src/template-part/edit/utils/hooks.js#L95) to learn how to convert blocks to template part and update the attributes.

5. After calling the `updateBlockAttributes` action, the save button should be clickable, and then clicking the save button will apply the changes, but if not, please check/update the `savehub` logic [here](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce-admin/client/customize-store/assembler-hub/sidebar/save-hub.tsx#L132-L168).

7. When modifying but not saving and returning to the main page, the changes should be reverted. Please refer to [these lines](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce-admin/client/customize-store/assembler-hub/sidebar/save-hub.tsx#L92-L131) if the logic need updates.

Gutenberg site editor header replacement reference:

https://github.com/woocommerce/woocommerce/assets/4344253/7514711b-4514-40c3-ab6d-8a5ae0732391

## Acceptance Criteria

```
Scenario: Header panel is opened
    Given the user opens the header panel
    Then available header patterns are shown

Scenario: User changes the header 
    Given the changes the header
    Then the header should be updated in the preview pane

Scenario: Save
    Given the user presses the save button
    Then the rendered preview should be saved permanently

```

### PR DESCRIPTION
**Pull request by samueljseay**
### Changes

This PR closes #39718 by loading header patterns in the header customization menu of the assembler.

I've also made some structural changes to move the `EntityProvider` higher up into the layout. I've done this because it will allow the side navigation to share access to the blocks being previewed in `BlockPreview`. I also added a helper hook `useEditorBlocks` so that it's simple for other components to gain access to the blocks and manipulate them.

### Preview

https://github.com/woocommerce/woocommerce/assets/1281828/8b265380-eb98-406a-9c05-76b14c324715

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

Add header customization to the Assembler Hub

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce-admin/webpack.config.js [hunk 1]
```diff
@@ -169,7 +169,7 @@ const webpackConfig = {
 	plugins: [
 		...styleConfig.plugins,
 		// Runs TypeScript type checker on a separate process.
-		! process.env.STORYBOOK && new ForkTsCheckerWebpackPlugin(),
+		// ! process.env.STORYBOOK && new ForkTsCheckerWebpackPlugin(),
```
Comment (samueljseay) [inline plugins/woocommerce-admin/webpack.config.js]: Ignore, just me getting frustrated by TS errors while trying to develop :P I'll remove this when it goes out of draft.

File: plugins/woocommerce-admin/client/customize-store/assembler-hub/layout.tsx [hunk 2]
```diff
@@ -66,119 +68,136 @@ export const Layout = () => {
...
 			value={ {
 				logoBlock,
 				setLogoBlock,
 			} }
 		>
-			<div className={ classnames( 'edit-site-layout' ) }>
-				<motion.div
-					className="edit-site-layout__header-container"
-					animate={ 'view' }
+			<EntityProvider kind="root" type="site">
```
Comment (samueljseay) [inline plugins/woocommerce-admin/client/customize-store/assembler-hub/layout.tsx]: This diff is hard to read but all I did was move EntityProvider here so we can access blocks from inside side nav

File: plugins/woocommerce-admin/client/customize-store/assembler-hub/sidebar/sidebar-navigation-screen-header.tsx [hunk 3]
```diff
@@ -30,9 +66,30 @@ export const SidebarNavigationScreenHeader = () => {
...
+								showTitlesAsTooltip={ true }
+							/>
+						) }
+					</div>
 				</>
 			}
 		/>
 	);
+
+	// return null;
```
Comment (chihsuan) [inline plugins/woocommerce-admin/client/customize-store/assembler-hub/sidebar/sidebar-navigation-screen-header.tsx]:
```suggestion

```

**Review by chihsuan — CHANGES_REQUESTED** [review]
Hey @samueljseay, Thanks for adding this super quickly. 👍 

I just noticed some issues:

1. It seems that the header patterns cannot be selected again when reloading the page after saving. Could you take a look?
https://github.com/woocommerce/woocommerce/assets/4344253/d7ad4f42-57d9-476e-bf82-6a445351c4d2

2. We should add a focus style border to the selected header pattern on the sidebar. Also need to start with the opacity preview frame (except the header) — when the user scrolls down, bring it back to 100%.

Header

3. We only need to show these four header patterns (Sorry, didn't have it in the issue):

Slug:
- woocommerce-blocks/header-centered-menu-with-search
- woocommerce-blocks/header-essential
- woocommerce-blocks/header-large
- woocommerce-blocks/header-minimal

**Review by moon0326 — COMMENTED** [review]
Thank you for the changes :) This is a good reference PR for me.

While testing the changes, I've noticed a possible bug in an edge case.

Steps to reproduce

1. Go to the header page
2. Refresh
3. Click `Assembler Hub`
4. Notice the blocks render twice.

https://github.com/woocommerce/woocommerce/assets/4723145/41c538eb-d101-4719-a717-a7c02fe12217

**Comment by samueljseay** [comment]
@moon0326 I explored the flickering. As you mentioned it happens if you page refresh and it auto navigates to this screen. unfortunately I think this is some bug with the rendering of the layout and navigation screen code that causes redundant re-renders of `SidebarNavigationScreenHeader`, which in turn re-renders the `BlockPatternList`. I haven't got to the bottom of this yet, but I'm thinking maybe it needs to be solved in a separate PR.

you can see it happen on the color pallette screen too for example

**Comment by samueljseay** [comment]
@chihsuan I've fixed the issue with saving, I've added some code to handle opacity too, but do you have a suggestion for overriding styles for items in `BlockPatternList`? kinda feels like we need something like emotion to add some targeted styles, otherwise I can add some global styles to fix it.

**Review by chihsuan — APPROVED** [review]
> I've fixed the issue with saving, I've added some code to handle opacity too, but do you have a suggestion for overriding styles for items in BlockPatternList? kinda feels like we need something like emotion to add some targeted styles, otherwise I can add some global styles to fix it.

Hey @samueljseay Thanks for the fix! I confirm that I can save changes successfully 👍  I think adding global styles is probably the only way at the moment. We could file an issue on Gutenberg repo to support that. 😄 

There seem to be a few issues but I think we can address as a follow-up to unblock other tasks as discussed:

1. After jumping out of the header sidebar menu, opacity still exists. We should remove it. 🙂 

https://github.com/woocommerce/woocommerce/assets/4344253/5f7a83dc-3f84-4719-945e-2576ee212cf4

2. I haven't been able to reliably reproduce this issue yet, but it seems that there are some things that may cause the page to become unresponsive. 🤔 

Screenshot 2023-09-14 at 10 27 25

**Comment by samueljseay** [comment]
> After jumping out of the header sidebar menu, opacity still exists. We should remove it. 🙂

😅 good point, This wasn't hard so I added it.

> I haven't been able to reliably reproduce this issue yet, but it seems that there are some things that may cause the page to become unresponsive. 🤔

Ok I haven't seen this yet but I'll keep an eye on it 👁️

**Pull request merged**